Search page

Emmanuel-Develops opened the pull request:
- [x] search api
- [x] use search
- [x] sidebar with filter and sort
- [x] result cards
- [x] toggle pills in result card
- [x] link pills from explore section to search
- [x] mobile responsiveness
- [x] use server action for search call to better integrate with [search API](https://github.com/bitcoinsearch/bitcoinsearch-app/pull/155)

0tuedon commented:
Asides the niit from the call? ready for review? @Emmanuel-Develops

Emmanuel-Develops commented:
> Asides the niit from the call? ready for review? @Emmanuel-Develops

core functionality, yes.
There's still some ui/ux nits to work out

IgboPharaoh commented:
Hello @Emmanuel-Develops can you please rebase your branch to the latest on staging to resolve merge conflicts

kouloumos commented on src/components/search/SearchResultCard.tsx:
```diff
@@ -0,0 +1,43 @@
+}
+
+export const SearchResultCard = ({ result, className }: { result: Result, className?: string }) => {
+  const { transcript_source, title, body, authors, tags, id, indexed_at: date, url } = result;
```
You should use `created_at` and not `indexed_at` for date

kouloumos commented on src/config/index.ts:
```diff
@@ -1 +1,45 @@
+
+export const setup = {
+  facetFields: ["authors", "transcript_source", "tags"] as FacetKeys[],
+  sortFields: ["indexed_at"]
```
same here, `created_at` not `indexed_at`

kouloumos reviewed (commented):
The implementation looks good overall, and I was able to test it locally. Search functionality works as expected.

However, I noticed a few issues and potential areas for improvement:

### Testing Prerequisites

To test locally, you need to define the `API_KEY` and `CLOUD_ID` environment variables. This requirement should be mentioned in the README, similar to how we handle it for bitcoin-search.

### Code Reuse

For files copied fully or partially from bitcoin-search, it would be helpful to include a note at the top of the file with a reference link. This adds context for future developers.

Additionally, if there's significant duplicated logic, we should consider centralizing it. With TLDR coming (which will also have a search page), maintaining similar functionality across three repositories is not sustainable.

## Observations

A couple of issues I observed:

### Topics Filtering

search query `/search?search=dlcs sir` 

I see duplicate values for the very first tags, first tag filters correctly, the other displays a 404 page in the search results.

![issue-duplicate-tags](https://github.com/user-attachments/assets/6a5ed65a-3cc3-4bc4-9eb0-18d87019a7ec)

There seems to be an issue with keeping state in the searchbox when using search

![issue-with-filter-search](https://github.com/user-attachments/assets/f8fd7e86-ac31-4c62-adcf-3a5c28c3de35)

![issue-with-state](https://github.com/user-attachments/assets/507aec71-f8be-4ab4-8a4e-06b7e02a954f)

Those issues don't exist in the Speakers and Sources filters.

### Red colors

Is the red text for "No matching options" and the red "X" for the search part in the designs? They look out of place.

![image](https://github.com/user-attachments/assets/48638b68-9427-4f63-bbd3-96f3b0e892f9)

![image](https://github.com/user-attachments/assets/4e5462e5-3e81-4a4e-900f-a64652b4cb06)

### Result Card

The date placement on the result card does not align with the design specifications and looks awkward.

![image](https://github.com/user-attachments/assets/7cf7f3b2-b51d-41e8-9a56-4c0ff1220fc1)

We should make a note for the source on the card. I know that is not an easy fix, but the way that is being displayed (slug instead of name) is not correct.

### Pagination isn't centered

![image](https://github.com/user-attachments/assets/ade082a4-dfb2-4419-bcb3-89a8ce304d4c)

![image](https://github.com/user-attachments/assets/f1550330-edde-4d82-b2f4-8443b240817e)

Emmanuel-Develops commented:
> For files copied fully or partially from bitcoin-search, it would be helpful to include a note at the top of the file with a reference link. This adds context for future developers.

Code that can be said to be partially duplicated are `useURLManager` and `useSearch`, and these hooks are self-contained functions (containing some different variables and functions in this implementation). Linking them to bitcoin-search doesn't provide any extra context.

Only /search/api contains more nuanced search, and I do not include it as it will be deleted during refactoring to use your global search API.
Notice in https://github.com/bitcointranscripts/registry/blob/search-page/src/app/search/searchCall.ts#L24 it can make a call to a provided URL rather than `/search/api` all custom data fetching will be deleted in favor of the global search API endpoint

also the envs won't be required anymore

Emmanuel-Develops commented:
> I see duplicate values for the very first tags, first tag filters correctly, the other displays a 404 page in the search results.

I am not getting the same results as you are (can't reproduce)
<img width="1181" alt="Screenshot 2025-01-14 at 11 28 41 AM" src="https://github.com/user-attachments/assets/958e41f9-f6c8-41d4-a1a0-cbe49257cc4a" />

> Is the red text for "No matching options" and the red "X" for the search part in the designs? They look out of place.

Mmm, there are no designs for this. Let's discuss that in the dev call today.
I'll switch it to grey in the meantime, but this isn't a blocker 👍

> The date placement on the result card does not align with the design specifications and looks awkward.

fixed!

> Pagination isn't centered

fixed!

cc: @kouloumos

kouloumos reviewed (approved):
Amazing! All my previous comments have been covered. Additionally, all of the places where "slug" was used, we are now using the full names.

This is the last one to be mapped, in the "Applied Filters":
![image](https://github.com/user-attachments/assets/2eb30a69-6e2d-4936-8953-7d06a5b5729d)

I think that it should be a quick fix by passing the selected filter from the same mapping function.

ACK as soon as the above is fixed.

Emmanuel-Develops commented:
> Amazing! All my previous comments have been covered. Additionally, all of the places where "slug" was used, we are now using the full names.
> 
> This is the last one to be mapped, in the "Applied Filters": ![image](https://private-user-images.githubusercontent.com/18506343/403435344-2eb30a69-6e2d-4936-8953-7d06a5b5729d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY5NDg4NzgsIm5iZiI6MTczNjk0ODU3OCwicGF0aCI6Ii8xODUwNjM0My80MDM0MzUzNDQtMmViMzBhNjktNmUyZC00OTM2LTg5NTMtN2QwNmE1YjU3MjlkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE1VDEzNDI1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhM2YyN2NlMGQ2OWNlZTUwYjMxNmY2MjU4Y2ZmODkwMWZhOGY5MjA0OTQ5YjZlMThlY2FlODdmY2IzMzQ5ZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Fm2UZqhoSmpul-J3bPln-kP0zgHpUPnu72PFc2XVlXQ)
> 
> I think that it should be a quick fix by passing the selected filter from the same mapping function.
> 
> ACK as soon as the above is fixed.

thanks, pushed the fix

kouloumos commented:
Looks good to me! :rocket:
